Caltech-IPAC/firefly
# Firefly 1193 ObsCore Packager

#### [Firefly-1193](https://jira.ipac.caltech.edu/browse/FIREFLY-1193):
- ObsCore Packager (similar to Sofia) 
- The naming convention I have used is: 
   - if it's a non-datalink file, get the filename from the access_url 
   -  Prepend the file with the product title template entries, or if that's not available, prepend it with '`${obs_collection}-${instrument_name}-${obs_id}`' 
   - For datalink files, I have created a folder with the name '`${obs_collection}-${instrument_name}-${obs_id}`' (or productTitleTemplate) 
      - and we get filename from access_url for every file in the datalink table that we want to download
- _Todo_: I need to finish implementing the logic to use the product title template (which comes from `getAppOptions().tapObsCore?.productTitleTemplate`) and I am sending it to the server via the search request, just need to finish writing the logic for that.  
   - Once this is done, productTitleTemplate will describe what should prepend the filename (or in the case of datalink, what should be the name of the folder)  
- _Todo_: I haven't accounted for duplicate file names. The user's OS might append (1), (2), etc. to duplicate file names but we should take care of it on our end, right? 
- _Todo_: minor, but I can add a check so that if there's only 1 file that we want to download from a datalink table, don't create a folder for it 
- **_Issue_**: I have set the BaseFileName to "ObsCore_Files", which what you should see in the DownloadOptionPanel as well. But when the file is download, it's downloaded as "ObsCore_Files_part1"...I am not sure where "part1" comes from, looking into it. 

**Updates 4/12**
- PR ready
- addressed feedback below from @loitly and wrote file naming code based on discussions with @robyww and @loitly 
- I've added comments through the code to better explain what different parts do (may remove some if there's too many), let me know if you have any questions 
- The logic that was used for file naming was (the getFileName function): 
   - if a productTitleTemplate is available from the client side, use it
   - else, use obs_title to name the file if it's available 
   - else, create a manual template (based on obs_collection, instrument_name and obs_id) 
- added checks for access_url and access_format/content_type (to get file extension) being null 
- BaseFileName (the zip folder name) is also changed to `tbl_id_Files` 
- test builds and instructions updated below 

**Updates 4/13**
- addressed PR comments, highlights:
  - for both file and folder names (and also the table title - which is the name of the downloaded zip folder), got rid of spaces and non-(alphanumeric,period,hyphen,underscore) characters and replaced the latter with either hyphen or underscore as appropriate (hyphen between column names, and underscore for any weird characters within a column name) 
  - creating folder for datalink table only if it has more than one valid file 
  - other small changes and bug fixes 
- _**Potential issue**_:  for some entries, I was getting this error msg in the Data Product viewer: 
    - `No displayable data available for this row: Request failed with status 500: https://fireflydev.ipac.caltech.edu/firefly-1193-package-obscore/firefly/CmdSrv/sync?cmd=upload`
    - to reproduce this error, search for TAP -> CADC -> m1, and under obs_id, search for 2827296
    - also, if you try and download this datalink entry, it succeeds but creates a folder even though there's only one file in the folder. The reason is that the datalink table still has more than 1 valid files (2 `previews` and 1 `this` specifically), but there was an error encountered in packaging these 3 files (see the README file)
    - I am not sure if this is an expected error or not, but wanted to highlight it. 
- both test builds below are updated with @robyww's  TAP UI changes as well

**Updates 4/14**
  - removed checks for uniqueness for file names, as @robyww's changes take care of that now 
  - only added checks for datalink folder name's uniqueness by calling FileUtil's `getUniqueFileNameForGroup` function
  - for non-datalink files, using a temp file to call URLDownload and using this temp file in the packager (instead of using the URL twice) 
     - for datalink files, we still need the access_url when creating the final FileInfo object to be packaged, but this only happens once per datalink entry anyway  

#### Testing:  
**Firefly**: https://fireflydev.ipac.caltech.edu/firefly-1193-package-obscore/firefly
**Irsaviewer** **(DCE)**: https://firefly-1193-package-obscore.irsakudev.ipac.caltech.edu/irsaviewer/dce
- Search for something that returns at least some datalink entries. 
    - On firefly, from the TAP panel, select CADC as TAP service, search for m1 (set row limit to 5 or 10 for faster results).
    -  On DCE, you may click on the MSX collection, click on a point (or, use this: 2h44m14.43s +57d57m24.8s Equ J2000) and search
- Click on the "Prepare Download" button and attempt to download single or multiple files. For datalink tables with more than one valid file that we want to download, a sub-folder will be created for this datalink table. 
- attempt to download non-datalink files as well (jpeg, fits,etc) directly 
- **Testing uniqueness when it comes to naming files**
   - to test that file names are always unique, download a datalink table that has at least 2 "**preview**" entries (in the **semantics** column)
      - for this, in your datakink folder, you should get file names, for instance, that look like this: `obsCollection-instrumentName-obsID-preview.jpeg` and `obsCollection-instrumentName-obsID-preview-1.jpeg `
